src/dynamic_boundary_conditions/tide_data.py

```diff
@@ -1,148 +1,449 @@
+        An integer representing the number of days before the highest tide to extract data for.
+        Must be a positive integer.
+    days_after_peak : int
+        An integer representing the number of days after the highest tide to extract data for.
+        Must be a positive integer.
+    """
+    # Check for invalid arguments
+    if days_before_peak < 0:
+        raise ValueError("'days_before_peak' must be at least 0.")
+    if days_after_peak < 0:
+        raise ValueError("'days_after_peak' must be at least 0.")
+    #  Get the datetime of the most recent highest tide that occurred within the requested time period.
+    highest_tide_datetime = get_highest_tide_datetime(tide_data)
+    # Get the start_date (first day before peak) and the end_date (last day after peak)
+    start_date = highest_tide_datetime.date() - timedelta(days=days_before_peak)
+    end_date = highest_tide_datetime.date() + timedelta(days=days_after_peak)
+    # Get a list of dates from start_date (first day before peak) to end_date (last day after peak)
+    dates_list = pd.date_range(start_date, end_date, freq='D').date.tolist()
+    return dates_list
 
-    # Start date (can be in the past or present) for collection of data
-    # String for start date of data in format of ('yyyy-mm-dd')
-    start_date = date.today().isoformat()
 
-    # String for datum type.  LAT: Lowest astronomical tide; MSL: Mean sea level
-    datum = "LAT"
+def find_existing_and_missing_dates(
+        tide_data: pd.DataFrame,
+        dates_list: List[date]) -> Tuple[List[date], List[date]]:
+    """
+    Check whether each date in the given list of dates is present or missing in the 'datetime_nz' column of the
+    given tide_data DataFrame. Returns two lists - existing_dates and missing_dates.
+    'existing_dates' contains dates from 'dates_list' that are present in the 'tide_data' DataFrame and
+    'missing_dates' contains dates from 'dates_list' that are not present in the 'tide_data' DataFrame.
+    The tide data for the missing dates needs to be fetched from NIWA to obtain a complete set of required
+    tide data surrounding the highest tide.
 
-    # How many days either side of the highest tide do you want data for:
-    before_n_days = 2
-    after_n_days = 4
+    Parameters
+    ----------
+    tide_data : pd.DataFrame
+        The original tide data obtained for the requested period with the time column expressed in NZ timezone
+        (converted from UTC).
+    dates_list: List[date]
+        A list of dates for the requested time period surrounding the highest tide. This includes all dates from
+        the start date (the first day before the highest tide) to the end date (the last day after the highest tide).
+    """
+    tide_data_dates = set(tide_data['datetime_nz'].dt.date)
+    dates_set = set(dates_list)
+    existing_dates = sorted(list(tide_data_dates.intersection(dates_set)))
+    missing_dates = sorted(list(dates_set - tide_data_dates))
```

Comment:
```suggestion
    existing_dates = sorted(tide_data_dates.intersection(dates_set))
    missing_dates = sorted(dates_set - tide_data_dates)
```
sorted() already will produce a list

Reply:
done :)